Add a script to automate environment setup

settermjd opened the pull request:
This PR creates a script for automating the setup of ownCloud, right up to the point of either using the installation wizard or command line installer. It:

- Installs `wget`, `make`, `npm`, `nodejs`, `unzip`, & `git`
- Ensures nodejs is available in the user path, fixing it if necessary
- Runs `make` to install the PHP and JS dependencies

It's not, currently, special. But it's a good start. It was written after going through [the development environment setup documentation](https://doc.owncloud.com/server/latest/developer_manual/general/devenv.html) and seeing that there was both incorrect and missing information — and that a script might just save users a lot of time.

### Supports

- **Debian/Ubuntu:** 16.04
- **openSUSE Leap:** 42.3
- **CentOS:** 7

All constructive feedback welcome.

### Assumptions

- Apache (or another web server) are already installed

settermjd committed:
Add a script to automate environment setup

This is the first version of a script for automating the setup of
ownCloud, right up to the point of either using the installation wizard
or command line installer. It assumes that Apache (or another webserver)
are already installed.

PVince81 commented on bin/setup-environment.sh:
```diff
@@ -0,0 +1,122 @@
+function install_required_packages()
+{
+  # Ensure that Apt’s cache is up to date
+  sudo apt-get -y -q update
```
won't work on opensuse

settermjd replied:
That's right. That's what it says in the script, and in the PR description :-) It's for Debian-based distros only atm. Time allowing, I'll expand it out.

PVince81 commented:
While I understand the feeling and usefulness, I have mixed feeling about this, considering that it will not work on all major Linux distributions.

tboerger commented on bin/setup-environment.sh:
```diff
@@ -0,0 +1,122 @@
+  if [ -z "$( which node )" ]; then
+    echo 'Node is installed, but not correctly set in the system path.'
+    echo 'Linking ' "$nodejs" ' to /usr/bin/node'
+    sudo ln -s "$nodejs" /usr/bin/node
```
Ubuntu and Debian should provide a node legacy package for that, no need to manually symlink

settermjd replied:
Thanks for the pointer.

tboerger commented:
Maybe we should push docker-based environments instead? I'm working on multiple dock-compose files within the core repo.

PVince81 commented:
If we can provide a dev friendly docker with everything already setup (including XDebug) this could be nice for people to quickly start.

settermjd commented:
I see what you're both saying about Docker. But while the documentation currently covers manual installation it makes sense to provide something like this — at least in the interim.

settermjd committed:
Replace manual link of nodejs with nodejs legacy package

mmattel commented:
Why not naming it ``setup-environment-debian.sh`` ?

PVince81 commented:
and then have `make setup` that detects your distro and runs the matching script...

settermjd commented:
Both are good points. Just about got it supporting openSUSE 42.3 atm.

settermjd committed:
Update the setup script to support openSUSE

mmattel commented on bin/setup-environment.sh:
```diff
@@ -0,0 +1,126 @@
+#!/bin/bash
+#
+## 
+## Usage: setup-environment.sh 
```
to be added in the usage description: ``<webserver user>`` ``<owncloud install dir>``

settermjd committed:
Added support for CentOS (7)

This, like openSUSE Leap, needed custom support for installing
PhantomJS. Otherwise, it was almost the same. A few changes, but quite
similar nonetheless.

settermjd committed:
Update usage note

settermjd commented:
Support for openSUSE and CentOS is now available.

mmattel commented on bin/setup-environment.sh:
```diff
@@ -0,0 +1,166 @@
+IFS=$'\n\t'
+REQUIRED_ARGS=2
+
+if [ "$( whoami )" == 'root' ]
```
do I get it right that this ``whoami`` code block is double present and not necessary because of ``function check_user()`` ?

settermjd replied:
Yes, you are. That must have been left by mistake from the original version of the script.

tboerger commented on bin/setup-environment.sh:
```diff
@@ -0,0 +1,166 @@
+
+function which_distro()
+{
+  case "$( grep -Eoi 'Debian|SUSE|Ubuntu' /etc/issue )" in 
```
I think it's more common to parse the `/etc/os-release`, which already provides variables if you eval the file.

settermjd replied:
Ah great. Thanks for that. I'd not heard of that file before, but just had a quick skim through https://www.freedesktop.org/software/systemd/man/os-release.html. So I can see that using it is a much better way of doing what I was attempting to do.

settermjd committed:
Remove unnecessary double user check

The extra user check is left-over from an earlier version of the script
before functions were introduced.

settermjd committed:
Update the distro detection function

Thanks to @tboerger for the suggestion to use /etc/os-release to
retrieve information to determine the distribution and version, instead
of the previous approach. It makes the script more uniform and
maintainable in how it works.

settermjd commented:
@tboerger, thanks for the suggestion. Advice implemented.

tboerger commented on bin/setup-environment.sh:
```diff
@@ -0,0 +1,173 @@
+
+function which_distro()
+{
+  echo $( grep '^ID=' /etc/os-release ) | awk '{split($0,array,"=")} END { print array[2] }'
```
```
source /etc/os-release
echo $ID
```

tboerger commented on bin/setup-environment.sh:
```diff
@@ -0,0 +1,173 @@
+
+function distro_version()
+{
+  echo $( grep '^VERSION_ID=' /etc/os-release ) | awk '{split($0,array,"=")} END { print array[2] }' | tr -d '""'
```
```
source /etc/os-release
echo $VERSION_ID
```

DeepDiver1975 commented:
I honestly don't see this something to be part of core. Can we put this into the documentation somewhere or maybe post this in some blog or something?

This script needs constant development to keep up to date with the changes of distros. No idea who will do this.

If this script is part of core the expectation is that we keep this up to date.

I'm closing this for now. THX

mmattel commented:
@settermjd maybe something to be added to the documentation/development section ?
